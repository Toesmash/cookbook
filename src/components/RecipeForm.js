import React from 'react';
import * as Yup from 'yup';
import { withFormik, Form, Field } from 'formik';

export const RecipeForm = ({ errors, touched }) => {
  return (
    <div>
      <Form>
        <label>Title</label>
        <Field type="text" name="title" component="input" />
        {errors.title && touched.title && <div>{errors.title}</div>}

        <label>Portions</label>
        <Field type="number" name="feeds" component="input" />
        {errors.feeds && touched.feeds && <div>{errors.feeds}</div>}

        <label>Preparation time in minutes</label>
        <Field type="number" name="preparation" component="input" />
        {errors.preparation && touched.preparation && <div>{errors.preparation}</div>}

        <label>Ingredients</label>
        <Field type="text" name="ingredients" component="textarea" />
        {errors.ingredients && touched.ingredients && <div>{errors.ingredients}</div>}

        <label>Instructions</label>
        <Field type="text" name="instructions" component="textarea" />
        {errors.instructions && touched.instructions && <div>{errors.instructions}</div>}

        <button type='submit'>Submit</button>
      </Form>
    </div>
  );
}

const schema = {
  title: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)]*$/, 'No special characters allowed'),
  feeds: Yup.number().required('Required field').min(1, 'Minimum number of portions is 1').max(8, 'Maximum number of portions is 8'),
  preparation: Yup.number().required('Required field').min(10, 'Minimum preparation time is 10minutes').max(300, 'Maximum preparation time is 5 hours'),
  ingredients: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)\n\s]*$/, 'No special characters allowed'),
  instructions: Yup.string().required('Required field').matches(/^[a-zA-Z0-9,\.\-\/:\(\)\n\s]*$/, 'No special characters allowed')
};

export const FormikRecipeForm = withFormik({
  mapPropsToValues({ recipe }) {
    return {
      title: recipe ? recipe.title : '',
      feeds: recipe ? recipe.feeds : '',
      preparation: recipe ? recipe.preparation : '',
      ingredients: recipe ? recipe.ingredients.join('\n') : '',
      instructions: recipe ? recipe.instructions.join('\n') : ''
    }
  },
  validationSchema: Yup.object().shape(schema),
  handleSubmit(values, { props }) {
    props.onSubmit({
      ...values,
      ingredients: values.ingredients.split('\n').filter(item => item),
      instructions: values.instructions.split('\n').filter(item => item)
    });
  }

})(RecipeForm);

export default FormikRecipeForm;




//   onSubmit = (e) => {
//     e.preventDefault();
//     if (!this.state.title || !this.state.ingredients || !this.state.instructions) {
//       this.setState(() => ({ error: 'Please set correct' }))
//     }
//     else {
//       this.setState(() => ({
//         error: ''
//       }))
//       this.props.onSubmit({
//         title: this.state.title,
//         feeds: this.state.feeds,
//         preparation: this.state.preparation,
//         ingredients: this.state.ingredients.filter((item) => (item)),
//         instructions: this.state.instructions.filter((item) => (item))
//       })
//     }
//   }





// export default class RecipeForm extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       title: props.recipe ? props.recipe.title : '',
//       feeds: props.recipe ? props.recipe.feeds : '',
//       preparation: props.recipe ? props.recipe.preparation : '',
//       ingredients: props.recipe ? props.recipe.ingredients : [],
//       instructions: props.recipe ? props.recipe.instructions : [],
//       error: ''
//     };
//   }

//   onTitleChange = (e) => {
//     const title = e.target.value;
//     this.setState(() => ({ title }))
//   }
//   onFeedsChange = (e) => {
//     const feeds = e.target.value;
//     this.setState(() => ({ feeds }))

//   }
//   onPrepChange = (e) => {
//     const preparation = e.target.value;
//     this.setState(() => ({ preparation }))

//   }
//   onInstructionChange = (e) => {
//     const instructions = e.target.value;
//     this.setState(() => ({ instructions: instructions.split("\n") }))
//   }
//   onIngredientsChange = (e) => {
//     const ingredients = e.target.value;
//     this.setState(() => ({ ingredients: ingredients.split("\n") }))
//   }

//   onSubmit = (e) => {
//     e.preventDefault();
//     if (!this.state.title || !this.state.ingredients || !this.state.instructions) {
//       this.setState(() => ({ error: 'Please set correct' }))
//     }
//     else {
//       this.setState(() => ({
//         error: ''
//       }))
//       this.props.onSubmit({
//         title: this.state.title,
//         feeds: this.state.feeds,
//         preparation: this.state.preparation,
//         ingredients: this.state.ingredients.filter((item) => (item)),
//         instructions: this.state.instructions.filter((item) => (item))
//       })
//     }
//   }

//   render() {
//     return (
//       <div>
//         {this.state.error && <p>{this.state.error}</p>}
//         <form onSubmit={this.onSubmit}>
//           <input
//             autoFocus
//             type="text"
//             name="title"
//             placeholder="Title"
//             value={this.state.title}
//             onChange={this.onTitleChange}
//           />
//           <input
//             type="text"
//             name="feeds"
//             placeholder="2"
//             value={this.state.feeds}
//             onChange={this.onFeedsChange}
//           />
//           <input
//             type="text"
//             name="preparation"
//             placeholder="60"
//             value={this.state.preparation}
//             onChange={this.onPrepChange}
//           />
//           <textarea
//             placeholder="Ingredients"
//             value={this.state.ingredients.join("\n")}
//             onChange={this.onIngredientsChange}
//           />
//           <textarea
//             placeholder="Instructions"
//             value={this.state.instructions.join("\n")}
//             onChange={this.onInstructionChange}
//           />
//           <button>Submit</button>
//         </form>
//       </div>
//     );
//   }
// }


