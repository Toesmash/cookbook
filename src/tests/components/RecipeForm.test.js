import React from 'react';
import { shallow } from 'enzyme';
import RecipeForm from '../../components/RecipeForm';
import dummyData from '../../redux/fixtures/dummyData';

test('Render RecipeForm without recipe data', () => {
  const wrapper = shallow(<RecipeForm />);
  expect(wrapper).toMatchSnapshot();
});

test('Render RecipeForm with recipe data', () => {
  const wrapper = shallow(<RecipeForm recipe={dummyData[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('Render error when submitting incomplete RecipeForm', () => {
  const wrapper = shallow(<RecipeForm />);
  wrapper.find('form').simulate('submit', { preventDefault: () => { } });
  expect(wrapper).toMatchSnapshot();
});