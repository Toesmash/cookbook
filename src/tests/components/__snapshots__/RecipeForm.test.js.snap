// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render RecipeForm with recipe data 1`] = `
<Formik
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "feeds": 4,
      "ingredients": "1kg zemiaky
300g hladka muka
sol
250g bryndza
100g slaninka",
      "instructions": "Ocistit zemiaky, nastruhat, pridat muku a sol
Zamiesit cesto
Vlozit do vriacej vody
Zmiesat s bryndzou",
      "preparation": 60,
      "title": "halusky",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  recipe={
    Object {
      "feeds": 4,
      "id": "1",
      "ingredients": Array [
        "1kg zemiaky",
        "300g hladka muka",
        "sol",
        "250g bryndza",
        "100g slaninka",
      ],
      "instructions": Array [
        "Ocistit zemiaky, nastruhat, pridat muku a sol",
        "Zamiesit cesto",
        "Vlozit do vriacej vody",
        "Zmiesat s bryndzou",
      ],
      "preparation": 60,
      "rating": true,
      "title": "halusky",
    }
  }
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;

exports[`Render RecipeForm without recipe data 1`] = `
<Formik
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "feeds": "",
      "ingredients": "",
      "instructions": "",
      "preparation": "",
      "title": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;
